pkg/agent/manager/comm.go

```diff
@@ -0,0 +1,111 @@
+package manager
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"sync"
+
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/spire/proto/api/node"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+type client struct {
+	conn   *grpc.ClientConn
+	stream node.Node_FetchSVIDClient
+}
+
+type clientsPool struct {
+	// Map of client connections to the server keyed by SPIFFEID.
+	clients map[string]*client
+	// Protects access to the pool.
+	m *sync.Mutex
+}
+
+func (m *manager) newGRPCConn(svid *x509.Certificate, key *ecdsa.PrivateKey) (*grpc.ClientConn, error) {
+	var tlsCert []tls.Certificate
+	var tlsConfig *tls.Config
+
+	spiffePeer := &spiffe_tls.TLSPeer{
+		SpiffeIDs:  []string{m.serverSPIFFEID},
+		TrustRoots: m.bundleAsCertPool(),
+	}
+	tlsCert = append(tlsCert, tls.Certificate{Certificate: [][]byte{svid.Raw}, PrivateKey: key})
+	tlsConfig = spiffePeer.NewTLSConfig(tlsCert)
+	dialCreds := grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+
+	conn, err := grpc.Dial(m.serverAddr.String(), dialCreds)
+	if err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+// newClient adds a new client to the pool and associates it to the specified list of spiffeIDs.
+func (m *manager) newClient(spiffeIDs []string, svid *x509.Certificate, key *ecdsa.PrivateKey) error {
+	// If there is no pool yet, create one.
+	m.mtx.Lock()
+	if m.clients == nil {
+		m.clients = &clientsPool{clients: map[string]*client{}, m: &sync.Mutex{}}
+	}
+	m.mtx.Unlock()
+
+	conn, err := m.newGRPCConn(svid, key)
+	if err != nil {
+		return err
+	}
+
+	for _, id := range spiffeIDs {
```

Comment:
will there ever be more than one conn/client per spiffe id?

Reply:
Actually we have more than one spiffe id per connection for the agent's spiffe id and the server spiffe id. The idea is that the key corresponds to some parent spiffe id, then when we have to process a registration entry we took its parent spiffe id and use it to determine which client should be used to send the request to the server. Then if we have a reg entry which parent id is the server's spiffe id, we use the client created using the base svid and key, and also the same client should be used if the parent is the agent's spiffe id. Please let me know if I wasn't clear...or maybe I missed something.